tools/spawn/spawn.c

```diff
@@ -0,0 +1,414 @@
+		fprintf(stderr, "Invalid timeout value '%d'. Please specify a value in [0 ~ 60] seconds.\n",
+				g_opt.timeout);
+		return -1;
+	}
+
+	/* get command list from file */
+	cmd_fp = fopen(cmd_file, "r");
+	if (!cmd_fp) {
+		fprintf(stderr, "Could not open command file '%s' for reading\n", cmd_file);
+		exit(1);
+	}
+
+	while (!feof(cmd_fp)) {
+		char *line = NULL;
+		size_t len = 0, ret;
+
+		/* get command line */
+		ret = getline(&line, &len, cmd_fp);
+		if (ret < 0) {
+			if (line)
+				free(line);
+			break;
+		}
+		line[ret] = '\0';
+
+		/* chop endline */
+		if (strstr(line, "\r\n"))						/* windows format */
+			line[strlen(line) - 2] = '\0';
+		else if (line[strlen(line) - 1] == '\n')
+			line[strlen(line) - 1] = '\0';
+
+		/* check the length of line and comment */
+		if (strlen(line) == 0 || line[0] == '#') {
+			free(line);
+			continue;
+		}
+
+		/* add new command to list */
+		g_opt.cmd_list = (struct spawn_cmd *) realloc(g_opt.cmd_list,
+								(g_opt.num_of_cmds + 1) * sizeof(struct spawn_cmd));
+		if (!g_opt.cmd_list) {
+			fprintf(stderr, "Out of memory!\n");
+			free(line);
+			break;
+		}
+
+		memset(&g_opt.cmd_list[g_opt.num_of_cmds], 0, sizeof(struct spawn_cmd));
+		g_opt.cmd_list[g_opt.num_of_cmds++].cmd = line;
+	}
+
+	/* close file */
+	fclose(cmd_fp);
+
+	/* check count of commands */
+	if (g_opt.num_of_cmds == 0) {
+		fprintf(stderr, "Invalid command file '%s'\n", cmd_file);
+		exit(1);
+	}
+
+	/* get forks count */
+	g_opt.num_of_forks = strtoul(num_of_forks, NULL, 10);
+	if (g_opt.num_of_forks <= 0 || g_opt.num_of_forks > g_opt.num_of_cmds) {
+		fprintf(stderr, "The number of forks should be value in [1...%d]\n", g_opt.num_of_cmds);
+		exit(1);
+	}
+
+	return 0;
+}
+
+/*
+ * fork new process with command
+ */
+
+static int fork_cmd(struct fork_info *fi)
+{
+	pid_t pid;
+	int i;
+
+	char *cmd = NULL;
+
+	char *tok, **args = NULL;
+	int tok_count = 0;
+
+	int fd_null;
+	int ret = 1;
+
+	/* lock mutex */
+	pthread_mutex_lock(&g_opt.mt);
+
+	/* check if command to be forked is exist */
+	for (i = 0; i < g_opt.num_of_cmds; i++) {
+		if (!g_opt.cmd_list[i].status) {
+			cmd = strdup(g_opt.cmd_list[i].cmd);
+			fi->cmd = g_opt.cmd_list[i].cmd;
+			g_opt.cmd_list[i].status = 1;
+			break;
+		}
+	}
+
+	/* unlock mutex */
+	pthread_mutex_unlock(&g_opt.mt);
+
+	if (!cmd) {
+		fprintf(stderr, "No left command to be executed at thread[#%d].\n", fi->index);
+		return 0;
+	}
+
+	fprintf(stderr, "Running command '%s' at thread[#%d]\n", fi->cmd, fi->index);
+
+	/* parse command arguments */
+	tok = strtok(cmd, " ");
+	while (tok) {
+		/* allocate args */
+		args = realloc(args, (tok_count + 2) * sizeof(char *));
+		if (!args) {
+			fprintf(stderr, "Out of memory.\n");
+			free(cmd);
+
+			return -1;
+		}
+
+		/* set args */
+		args[tok_count] = tok;
+		tok_count++;
+
+		tok = strtok(NULL, " ");
+	}
+	args[tok_count] = NULL;
+
+	/* open /dev/null for redirecting output */
+	fd_null = open("/dev/null", O_WRONLY);
```

Comment:
I think this will leak a file descriptor in the parent process. Best open `/dev/null/` in the child (after the fork)